c++/src/kj/async-inl.h

```diff
@@ -377,13 +614,32 @@ class PromiseDisposer {
     PromiseArena* arena = next->arena;
 
     if (arena == nullptr ||
-        reinterpret_cast<byte*>(next.get()) - reinterpret_cast<byte*>(arena) < sizeof(T) ||
+        reinterpret_cast<byte*>(next.get()) - reinterpret_cast<byte*>(&arena->space) < sizeof(T) ||
         alignof(T) != alignof(void*)) {
       // No arena available, or not enough space, or weird alignment needed. Start new arena.
+      // TODO(perf): If we have an arena, it would be nice to propagate its EnvironmentSet somehow.
+      //   Perhaps a new alloc() overload?
       return alloc<T, D>(kj::mv(next), kj::fwd<Params>(params)...);
     } else {
       // Append to arena.
+
+#ifdef KJ_DEBUG
+      // In debug mode, we check that the `next` node's EnvironmentSet is the same as the current
+      // event loop's current environment set. This will alert the code author to mistakes such as:
+      //
+      //   Promise<void> promise1(READY_NOW);
+      //   Promise<void> promise2 = nullptr;
+      //   runInEnvironment([&]() {
+      //     promise1.then([]() {});  // ERROR: should have used .adoptEnvironment()
+      //     promise2 = READY_NOW;
+      //   });
+      //   promise2.then([]() {});  // ERROR: should have returned the promise
       //
+      // Unfortunately, we alert the code author to these mistakes by crashing, because
+      // `appendPromise()` is noexcept. Oh well.
+      requireEnvironmentSetEqual(next->getEnvironmentSet(), EnvironmentSet::tryGetCurrent());
```

Comment:
If we reduced this to an operator== (see comment on implementation), it might be cheap enough to do the check in release mode. And then, instead of throwing when it fails, we could actually add the missing node to the chain, maybe?